### ISSUE TITLE
Use of `typing.Self` in `__init__`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#8532]
### Discussed in https://github.com/microsoft/pyright/discussions/8531

<div type='discussions-op-text'>

<sup>Originally posted by **tpgillam** July 25, 2024</sup>
Consider the following example:

```python
import typing

class A:
    def __init__(self: typing.Self) -> None:
        pass
```

Pyright will flag a `reportInvalidTypeVarUse` over the use of `Self` here. However, by my reading of the recent specification, this (probably?) ought to be allowed:

https://github.com/python/typing/blob/b908de0205fb208cd2d74fe342ab33f28f35b71b/docs/spec/constructors.rst?plain=1#L206-L207

> If the ``self`` parameter within the ``__init__`` method is not annotated, type checkers should infer a type of ``Self``.

I am curious about what the correct resolution is for the inconsistency!

Also, in _practice_ I am guessing that the general consensus is to discourage explicit annotation of `self` in general? e.g. I note that Ruff's [ANN01](https://docs.astral.sh/ruff/rules/missing-type-self/) rule is deprecated.</div>

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
